### ISSUE TITLE
Vision for organizing tests

### ISSUE THREAD

**Issue by kyouko-taiga** [val-lang/val#436]
I want to open this issue to share my vision about the future of testing for Val compiler.

First, I want to distinguish between 3 categories of tests:
1. Unit tests of Swift components
2. Integration tests of compiler parts
3. End-to-end tests of compiled/interpreted Val programs.

Unit tests should obviously be written in Swift and organized according to the common practice. For integeration and end-to-end tests, I want to favor writing tests in Val rather than having Val fragments scattered across Swift files. I believe inline Val code should be used mostly for the purpose of unit testing components that require an AST.

Writing separate Val files will help us progressively build a corpus of programs that I believe to be valuable for a handful of reasons: 
1. to provide a collection of Val program examples that people can find easily.
2. to provide an easy way for people to contribute new tests without having to know anything about Swift.
3. to facilitate looking for recurring patterns in Val code with classical grep-like approaches.
4. to facilitate reusing test cases.

For the sake of simplicity, the proposed architecture currently assumes a program to be a single Val file. However, we should keep in mind that we'll eventually have to write tests for programs consisting of several files or even modules. That being said, I believe having a great setup for handling individual files is a clear step toward one that can handle multiple modules.

I want to avoid (without being completely opposed to) "hacking" features into Val for the purpose of testing it from Swift. Testing is an important part the developer's experience so we should be sure to develop a satisfying story in Val, and one that stands on its own. I believe having to eat our own dog food will help us making informed decisions.

Although building such a test framework in Val comes is quite ambitious, I believe we can do it progressively. We would start by simply running programs and check their exit status, quickly implement assertions to support more elaborate tests, and then slowly iterate on the design of a library.

Now let's have a look at the proposed arborescence of `Tests/`:

```
Tests/
  |- CompilerTests/
  |  |- CompilerTests.swift
  |  |- TypeChecking.testsuite/
  |  |- CXXCodeGen.testsuite/
  |  |- LLVMCodeGen.testsuite/
  |  |- Parsing.testsuite/
  |  |  |- UnterminatedComment.val
  |  |  |- ...
  |
  |- UnitTests/
  |  |- Core/
  |  |  |- SourceFileTests.swift
  |  |  |- ...
  |  |
  |  |- Utils/
  |     |- ...
  |
  |- Val
  |  |- ValTests.swift
  |  |- Foo.testsuite/
  |  |- Bar.testsuite/
  |  |- Library/
  |  |  |- Core.testsuite/
  |  |     |- IntTests.val
  |  |     |- ...
  |
  |- Utils/
  |  |- SomeUsefulComponent.swift
  |  |- ...
```

## `CompilerTests/`

This directory would contain integration tests.

A single test would consist of running one of the Val programs in one of the test suites. For example, `UnterminatedComment.val` would be a Val source file with test annotations, just like we have now.

`CompilerTests.swift` would implement the test runner for all of these tests, taking care of the integration with XCTest.

## `Val/`

This directory would contain end-to-end tests.

These tests would not use the annotations. Instead, they would be written as regular Val tests in the framework we'll be developing for Val, i.e. starting with exit statuses before moving to assertions and eventually something more sophisticated

Tests in this directory would be eventually executed for our three backends: C++, LLVM, and the interpreter. `ValTests.swift` would implement the test runners and take care of the integration with XCTest.

## `UnitTests/`

This directory would contain unit tests for the our individual Swift components. Each sub-directory would correspond to a target and contain a file for each tested component.

Nothing fancy here; just good old Swift unit tests.

## `Utils/`

This directory would contain the Swift helper components shared across multiple test targets, should we need some.

**Comment by lucteo** [comment]
This fits my purposes for near & medium terms.

While, ideally we should have more unit tests than integration tests, it feels to me that it would probably make more sense for us to have a larger corpus of integration tests. (at least in the areas I was working on)

I feel that the goals around "progressively build a corpus of programs" are not quite fitting the purpose of end-to-end tests. but that is something that we need to evaluate on the long run. So, I don't have material objections for the time being.

Bottom line, I'm on board, and eager to have more and more tests :)

**Comment by C-BJ** [comment]
Do we need to organize our tests this way now?
If so, I want to do a PR.

**Comment by kyouko-taiga** [comment]
Yes, I haven't had time to work on it yet. A PR will be greatly appreciated!

**Comment by C-BJ** [comment]
Assigned to me?

**Comment by C-BJ** [comment]
> Tests in this directory would be eventually executed for our three backends: C++, LLVM, and the interpreter.
What's the third backend?
@kyouko-taiga

**Comment by kyouko-taiga** [comment]
We originally planned to have an IR interpreter in addition to the LLVM and C++ backend. We have dropped this idea, though.

**Comment by peter-evans** [comment]
There are quite a few open issues related to tests, and I wondered if it might be a good idea to get the tests organized before attempting those other issues. Is there anything I can do to help get this over the line?

From the open PR:

> The PR gets us closer to what I'd like to implement, but we're not completely there yet. There are some changes we need to make on the tests themselves, not simply move files around.

Would it make sense to split this effort into stages? Perhaps merge the file location changes first, and then follow up afterwards with further changes to the tests themselves and how they are executed.

**Comment by kyouko-taiga** [comment]
The original document and the associated PR (#508) are way out of date. We can revive this topic but we would need a fresh start given the way our code base and approach to testing have evolved. The good news is that many things got simpler. We had plans for a C++ transpiler, an IR interpreter, _and_ an LLVM backend (yes, we were that ambitious!) Now we only have to care about the LLVM backend.

One thing about which I'm slowly changing my mind is whether we still need the three categories of tests I had identified:

> - Unit tests of Swift components
> - Integration tests of compiler parts
> - End-to-end tests of compiled/interpreted ~~Val~~ Hylo programs.

Back then it was obvious to me that we wanted to distinguish between parsing, type checking, and lowering tests. But now I'm not so sure anymore. In particular, the line between type checking and lowering is a little blurry as there are "type errors" that can only be reported after last use analysis runs on the IR (e.g., missing conformances to `Deinitializable`). Further, monomorphization (and soon existentialization) also interacts with type checking in ways that make it difficult to keep both phases strictly separated. So nowadays I tend to simply write end to end tests for everything.

Given this observation, we may want to consider dropping the integration tests of individual compiler parts and just write those as end to end tests, optionally with a flag saying how deep the test should go into the compilation pipeline.

One other thing on the horizon is that we'll need pretty is a way to test the import of individual modules. Currently all tests are one single source file defining a bunch of functions (including a `main`) and importing the standard library. But we certainly want to test projects that have more than one user source.

Finally, another aspect that wasn't part of the equation is that we now have standard library tests. Those will soon need to use some kind of testing framework and so they'll need a slightly different setup than the end to end tests.

I can spend some cycles on this problem but if you're interested @peter-evans that is a medium-sized project on which you could take the lead.
